changes for knative support

idurgakalyan opened the pull request:
<!--
PLEASE READ THIS MESSAGE.

Documentation fixes or enhancements:
- for Traefik v2: use branch v2.11
- for Traefik v3: use branch v3.2

Bug fixes:
- for Traefik v2: use branch v2.11
- for Traefik v3: use branch v3.2

Enhancements:
- for Traefik v2: we only accept bug fixes
- for Traefik v3: use branch master

HOW TO WRITE A GOOD PULL REQUEST? https://doc.traefik.io/traefik/contributing/submitting-pull-requests/

-->

### What does this PR do?

Add support for KnativeIngressRoute that can route calls to knative pods


### Motivation

knative is available for longtime and adding providing knative support to traefik can help manage all the routing with out using other components.


### More

- [ ] Added/updated tests
- [ ] Added/updated documentation

### Additional Notes

One the proposal is approved, will add test and update documentation further. The changes are validated with 1.31.0 version of kubernetes and 1.16.0 0f Knatives.

idurgakalyan commented:
Here is the Knative service and ingress route that are created.
-----
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: helloworld-go
  namespace: default
  annotations:
    serving.knative.dev/rollout-duration: "200s"
spec:
  template:
    metadata:
      annotations:
        networking.knative.dev/ingress-class: "traefik.ingress.networking.knative.dev"
    spec:
      containers:
        - image: ghcr.io/knative/helloworld-go:latest
          env:
            - name: TARGET
              value: "Go Sample v1"
  traffic:
  - percent: 60
    revisionName: helloworld-go-00001
    tag: prod
  - percent: 40
    revisionName: helloworld-go-00002
    tag: staging


----
apiVersion: traefik.io/v1alpha1
kind: IngressRouteKnative
metadata:
  name: hello-world-go
  namespace: default
spec:
  routes:
    - match: PathPrefix(`/hello-world-go`)
      kind: Rule
      services:
        - name: helloworld-go
          namespace: default

----
![Screenshot 2024-12-19 at 10 38 43 AM](https://github.com/user-attachments/assets/6e198f4c-e80f-48b3-a328-b0c1c294cb58)

idurgakalyan commented:
@nmengin Are you able to provide initial feedback before adding test cases and documentation?

nmengin commented:
Hello @idurgakalyan,

Thank you for your contribution.
We've added the PR on our plate to start an initial review.

As it's the end of the year most of our maintainers are not working, and they will have a busy January with the incoming version to release.

However, I can assure you that your PR is on our radar and we'll keep you updated following up on it.

kevinpollet commented:
Hello @idurgakalyan and thanks for your contribution,

From our understanding, to be a KNative Ingress Gateway, Traefik has to fulfill the [Knative Networking Specification](https://github.com/knative/networking) as explained in the [documentation](https://github.com/knative/networking). This means that the KNative provider has to watch and create a dynamic configuration for KIngress resources, as also explained [here](https://knative.dev/docs/serving/architecture/#traffic-flow-and-dns). In our case, the `net—*` controller would be a Traefik provider.

Are we missing anything here, or is there a specific reason to introduce a new `KnativeIngressRoute` resource?

In addition, we are exploring the possibility of introducing middleware support at the service or server level (which would allow the headers middleware to be used at the service level as required by KNative). As there is no consensus today, a possible solution could be to introduce the headers feature privately, but this needs to be discussed with other maintainers.

idurgakalyan commented:
Thanks for the review @kevinpollet. 

We are indeed watching for changes to Knative ingress resources. https://github.com/traefik/traefik/pull/11356/files#diff-090ad1ae4a224c8b897f4a6a4c34cbc6a8755127eec159d9ff80e735f43265a3R175-R176. I have modified the resource as part of validation to ensure new weights are reflected and requests are routed appropriately.

The reason for introducing KnativeIngressRoute resource specifically is to bring traefik rick features to Knative resources. Currently we are listening to changes in both KnativeIngressRoute and Knative ingress resources, if we determine we don't need KnativeIngressRoute, I can make the change to remove it.

Knative requires passing headers for appropriate routing, so I have added support for headers in WRR. I also assume that adding headers for weighted round-robin can support other use cases in the future.
https://github.com/traefik/traefik/pull/11356/files#diff-6e97fd86ea234a5fc4a9fefff3082dcff2f710fe55d28d56b5a5c32847c050baR21

The reason for middleware is to support tag based routing that is a functionality of Knative. https://knative.dev/docs/serving/traffic-management/.

If a tag is provided on Knative ingress resource, then it creates a new route that requires users to provide an additional header `revision-tag` to support routing to specific tag.

Please let me know if there are any other questions or suggestions.

idurgakalyan commented:
> In our case, the `net—*` controller would be a Traefik provider.

We are doing the same thing here https://github.com/traefik/traefik/pull/11356/files#diff-23ba4b70f2b27b59d6683aa441c198b2726dd6f1d283e5982e7dcfd78cdb990cR56. For the given Knative Ingress we are are creating a router https://github.com/traefik/traefik/pull/11356/files#diff-23ba4b70f2b27b59d6683aa441c198b2726dd6f1d283e5982e7dcfd78cdb990cR102

kevinpollet commented:
Thanks for the answers @idurgakalyan,

> The reason for introducing KnativeIngressRoute resource specifically is to bring traefik rick features to Knative resources. Currently we are listening to changes in both KnativeIngressRoute and Knative ingress resources, if we determine we don't need KnativeIngressRoute, I can make the change to remove it.

Which traefik features would be needed to be supported? If we understand correctly the KNative Ingress is something created by the KNative automation/controller, so supporting only KNative resources is the way to go.

Furthermore, did you notice that they are providing conformance tests?
See https://github.com/knative/networking/blob/main/test/conformance/ingress/README.md
We think that it would be great to look at it to have an idea of our current support.

idurgakalyan commented:
Thanks @kevinpollet. 

I introduced KnativeIngressRoute similar to IngressRoute for Kubernetes provider. It supports rules, middleware, observability etc. I can remove KnativeIngressRoute and just create a router entry based on Knative Ingress controller. Let me know if thats the way we want to proceed with? 
 
WRT conformance tests, Are we expecting just validation or adding them for traefik as well?

idurgakalyan commented:
![Screenshot 2025-01-13 at 2 08 20 PM](https://github.com/user-attachments/assets/0a547b47-be3c-4e13-ac13-8c2919360767)
For each resource with a specified domain, knative shows two sets of hosts. One with cluster.local and other with domain example.com(custom domain). Would the recommend path be to ignore  cluster.local hosts?

kevinpollet commented:
Hello @idurgakalyan,

> I introduced KnativeIngressRoute similar to IngressRoute for Kubernetes provider. Knative doesn't support path based routing, it only supports sub-domains. Introducing KnativeIngressRoute supports entry points, rules, middleware, observability, path based routing etc. I can remove KnativeIngressRoute and just create a router entry based on Knative Ingress controller. Let me know if thats the way we want to proceed with?

Maybe I missed something but from my understanding, KNative ingresses are created by the automation and not by the user. The user only cares about its app and the corresponding service manifest. Thus, I am not sure that supporting entry points, rules, middleware, observability, through a new `KnativeIngressRoute` resources is something needed by the end user for the moment.

> WRT conformance tests, Are we expecting just validation or adding them for traefik as well?

I think it could be great to setup them in Traefik, and to be able to run them with a flag as this is done for the Gateway API conformance tests (e.g. https://github.com/traefik/traefik/blob/master/Makefile#L102).

> For each resource with a specified domain, knative shows two sets of hosts. One with cluster.local and other with domain example.com(custom domain). Would the recommend path be to ignore Visibility ClusterLocal?

I am not sure to follow, can you elaborate on this?

idurgakalyan commented:
@kevinpollet rebase had some issue created a new PR 
https://github.com/traefik/traefik/pull/11448/files

This PR now only validates on KnativeIngress. Once created it creates a router with the hosts defined in the KNativeIngress. Currently there is no way to specify entry point hence all the entry points are populated. Would this be expected?
<img width="1422" alt="Screenshot 2025-01-15 at 12 19 28 AM" src="https://github.com/user-attachments/assets/c3e82410-c094-4942-b228-9e90553f3364" />

In addition, KNativeService creates the following KNativeIngress for the following manifest(if custom domain configured)
```
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: helloworld-go
  namespace: default
spec:
  template:
    metadata:
      annotations:
        networking.knative.dev/ingress-class: "traefik.ingress.networking.knative.dev"
    spec:
      containers:
        - image: ghcr.io/knative/helloworld-go:latest
          env:
            - name: TARGET
              value: "Go Sample v1"
```
KNative Ingress:
```
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: helloworld-go
  namespace: default
  annotations:
    serving.knative.dev/rollout-duration: "200s"
spec:
  template:
    metadata:
      annotations:
        networking.knative.dev/ingress-class: "traefik.ingress.networking.knative.dev"
    spec:
      containers:
        - image: ghcr.io/knative/helloworld-go:latest
          env:
            - name: TARGET
              value: "Go Sample v1"
kinamp@APWPQWR92A1039 knative % kubectl describe ingresses.networking.internal.knative.dev helloworld-go
Name:         helloworld-go
Namespace:    default
Labels:       serving.knative.dev/route=helloworld-go
              serving.knative.dev/routeNamespace=default
              serving.knative.dev/service=helloworld-go
Annotations:  networking.internal.knative.dev/rollout:
                {"configurations":[{"configurationName":"helloworld-go","percent":100,"revisions":[{"revisionName":"helloworld-go-00001","percent":100}],"...
              networking.knative.dev/ingress.class: traefik.ingress.networking.knative.dev
              serving.knative.dev/creator: minikube-user
              serving.knative.dev/lastModifier: minikube-user
              serving.knative.dev/rollout-duration: 200s
API Version:  networking.internal.knative.dev/v1alpha1
Kind:         Ingress
Metadata:
  Creation Timestamp:  2025-01-15T06:48:28Z
  Generation:          1
  Managed Fields:
    API Version:  networking.internal.knative.dev/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:networking.internal.knative.dev/rollout:
          f:networking.knative.dev/ingress.class:
          f:serving.knative.dev/creator:
          f:serving.knative.dev/lastModifier:
          f:serving.knative.dev/rollout-duration:
        f:labels:
          .:
          f:serving.knative.dev/route:
          f:serving.knative.dev/routeNamespace:
          f:serving.knative.dev/service:
        f:ownerReferences:
          .:
          k:{"uid":"d07f6ee6-1784-4415-8ead-19b90cd85c96"}:
      f:spec:
        .:
        f:httpOption:
        f:rules:
    Manager:      controller
    Operation:    Update
    Time:         2025-01-15T06:48:28Z
    API Version:  networking.internal.knative.dev/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        .:
        f:conditions:
        f:observedGeneration:
        f:privateLoadBalancer:
          .:
          f:ingress:
        f:publicLoadBalancer:
          .:
          f:ingress:
    Manager:      traefik
    Operation:    Update
    Subresource:  status
    Time:         2025-01-15T06:48:28Z
  Owner References:
    API Version:           serving.knative.dev/v1
    Block Owner Deletion:  true
    Controller:            true
    Kind:                  Route
    Name:                  helloworld-go
    UID:                   d07f6ee6-1784-4415-8ead-19b90cd85c96
  Resource Version:        994114
  UID:                     b935a8b4-4164-49b8-a254-5400cbe67a0c
Spec:
  Http Option:  Enabled
  Rules:
    Hosts:
      helloworld-go.default
      helloworld-go.default.svc
      helloworld-go.default.svc.cluster.local
    Http:
      Paths:
        Splits:
          Append Headers:
            Knative - Serving - Namespace:  default
            Knative - Serving - Revision:   helloworld-go-00001
          Percent:                          100
          Service Name:                     helloworld-go-00001
          Service Namespace:                default
          Service Port:                     80
    Visibility:                             ClusterLocal
    Hosts:
      helloworld-go.default.example.com
    Http:
      Paths:
        Splits:
          Append Headers:
            Knative - Serving - Namespace:  default
            Knative - Serving - Revision:   helloworld-go-00001
          Percent:                          100
          Service Name:                     helloworld-go-00001
          Service Namespace:                default
          Service Port:                     80
    Visibility:                             ExternalIP
Status:
  Conditions:
    Last Transition Time:  2025-01-15T06:48:28Z
    Status:                True
    Type:                  LoadBalancerReady
    Last Transition Time:  2025-01-15T06:48:28Z
    Status:                True
    Type:                  NetworkConfigured
    Last Transition Time:  2025-01-15T06:48:28Z
    Status:                True
    Type:                  Ready
  Observed Generation:     1
  Private Load Balancer:
    Ingress:
  Public Load Balancer:
    Ingress:
Events:  <none>
```

There are two sets of hosts defined 
```
set1:
      helloworld-go.default
      helloworld-go.default.svc
      helloworld-go.default.svc.cluster.local
      ```
      ```
 Set2:
     helloworld-go.default.example.com
     ```
   In this case how do we want the routes to be defined? Should we be creating two routes one for custom domain and another for helloworld-go.default, helloworld-go.default.svc and helloworld-go.default.svc.cluster.local?

kevinpollet commented:
Superseded by https://github.com/traefik/traefik/pull/11448
